lib/cf/broker/src/test/oauth-test.js

```diff
@@ -1,74 +1,122 @@
 'use strict';
 
-const _ = require('underscore');
-const extend = _.extend;
 const abacusOauth = require('abacus-oauth');
 
-let oauth = require('../auth/oauth.js');
-
-// Mock abacus-debug
-const loggerSpy = spy((arg) => arg);
-require.cache[require.resolve('abacus-debug')].exports = spy(() => loggerSpy);
+const reloadOauthModule = () => {
+  delete require.cache[require.resolve('../auth/oauth.js')];
+  return require('../auth/oauth.js');
+};
 
 describe('Oauth', () => {
-  context('when reading credentials from the environment', () => {
-    it('should not fail when they are not set', () => {
-      delete process.env.SERVICE_BROKER_CLIENT_ID;
-      delete process.env.SERVICE_BROKER_CLIENT_SECRET;
-      expect(oauth.init).to.not.throw(Error);
-    });
-  });
+  const sandbox = sinon.sandbox.create();
 
-  context('when credentials are valid', () => {
-    const sampleToken = 'Bearer AAA';
+  const wrapperStub = sandbox.stub();
+  wrapperStub.start = sandbox.stub();
+  const oauthStub = sandbox.stub(abacusOauth, 'cache');
 
-    const mockAbacusOauth = (cacheResult, cacheStartCallbackResult) => {
-      const abacusOauthCacheSpy = spy(() => cacheResult);
-      abacusOauthCacheSpy.start = spy((cb) => cb(cacheStartCallbackResult));
-      const abacusOauthMock = extend({}, abacusOauth, {
-        cache: () => abacusOauthCacheSpy
-      });
-      require.cache[require.resolve('abacus-oauth')].exports = abacusOauthMock;
-      // reload oauth module to resolve the mock
-      delete require.cache[require.resolve('../auth/oauth.js')];
-      oauth = require('../auth/oauth.js');
-      return abacusOauthCacheSpy;
-    };
+  const client = 'client';
+  const secret = 'secret';
+
+  after(() => sandbox.restore());
+
+  afterEach(() => sandbox.reset());
+
+  context('when all credentials are provided', () => {
+    const scopes = 'scope1 scope2';
+    const token1 = 'Bearer AAA';
+    const token2 = 'Bearer BBB';
 
     beforeEach(() => {
-      process.env.SERVICE_BROKER_CLIENT_ID = 'valid';
-      process.env.SERVICE_BROKER_CLIENT_SECRET = 'secret';
+      process.env.SERVICE_BROKER_CLIENT_ID = client;
+      process.env.SERVICE_BROKER_CLIENT_SECRET = secret;
+      process.env.SERVICE_BROKER_CLIENT_SCOPES = scopes;
     });
 
-    it('should return an authorization header', () => {
-      const abacusOauthSpy = mockAbacusOauth(sampleToken, undefined);
+    it('should call abacus oauth with the corresponding arguments', () => {
+      oauthStub.returns(wrapperStub);
 
-      oauth.init(() => {
-        expect(oauth.authHeader())
-          .to.deep.equal({ authorization: sampleToken });
-        expect(abacusOauthSpy.called).to.equal(true);
-        expect(abacusOauthSpy.start.called).to.equal(true);
-        expect(loggerSpy.calledWithMatch('Successfully fetched'));
+      reloadOauthModule();
```

Comment:
Can we split this into two tests to avoid running reloadOauthModule in between?

Reply:
What do you mean to split it?

You're right that we've made a lot of reloads that are unnecessary, tests are now refactored  to make only one after the environment has changed.